mlflow/models/signature.py

```diff
@@ -224,24 +224,30 @@ def infer_signature(
     Returns:
       ModelSignature
     """
-    if model_input is not None:
-        inputs = (
-            convert_dataclass_to_schema(model_input)
-            if is_dataclass(model_input)
-            else _infer_schema(model_input)
-        )
-    else:
-        inputs = None
-    if model_output is not None:
-        outputs = (
-            convert_dataclass_to_schema(model_output)
-            if is_dataclass(model_output)
-            else _infer_schema(model_output)
-        )
-    else:
-        outputs = None
-    params = _infer_param_schema(params) if params else None
-    return ModelSignature(inputs, outputs, params)
+    schemas = {"inputs": model_input, "outputs": model_output}
+    for key, data in schemas.items():
+        if data is not None:
+            try:
+                schemas[key] = (
+                    convert_dataclass_to_schema(data) if is_dataclass(data) else _infer_schema(data)
```

Comment:
nit: does it make sense to put this logic in `_infer_schema()` instead? it seems like we can replace this block with just assigning to AnyType, so we might not have to raise an exception anymore (unless there are other situations when raising this exception is helpful):

https://github.com/mlflow/mlflow/blob/8747e6da6e7556c757372d008c9fbe17aaf15d9a/mlflow/types/utils.py#L415-L431

Reply:
just thinking that the code could look a bit simpler if so, something like

```
for key, data in schemas.items():
  if data is not None:
    schemas[key] = (
      convert_dataclass_to_schema(data) if is_dataclass(data) else _infer_schema(data)
    )

if schema["outputs"] == Schema([ColSpec(type=AnyType())]:
  # warn
```

Reply:
Emm it's possible other pieces also raise exception in _infer_schema, and I didn't want to wrap the whole function with a try catch

Reply:
makes sense!